whistl/train.py

```diff
@@ -0,0 +1,480 @@
+    IRM looks for features in the data that are invariant between different environments, as
+    they are more likely to be predictive of true causal signals as opposed to spurious
+    correlations. For more information, read https://arxiv.org/abs/1907.02893
+
+    Arguments
+    ---------
+    classifier: pytorch.nn.Module
+        The class of model to train
+    map_file: string or Path
+        The file created by label_samples.py to be used to match samples to labels
+    train_dirs: list of str
+        The directories containing training data
+    tune_dirs: list of str
+        The directories containing tuning data
+    gene_file: string or Path
+        The path to the file containing the list of genes to use in the model
+    num_epochs: int
+        The number of times the model should be trained on all the data
+    loss_scaling_factor: float
+        A hyperparameter that balances the classification loss penalty with the study invariance
+        penalty. A larger value of loss_scaling_factor will cause the loss to count for more and
+        the invariance penalty to count for less
+    label_to_encoding: dict
+        A dictionary mapping string labels like 'sepsis' to an encoded form like 0 or 1
+    device: torch.device
+        The device to train the model on (either a gpu or a cpu)
+    logger: logging.logger
+        The python logger object to handle printing logs
+    save_file: string or Path object
+        The file to save the model to. If save_file is None, the model won't be saved
+    burn_in_epochs: int
+        The number of epochs at the beginning of training to not save the model
+
+    Returns
+    -------
+    results: dict
+        A dictionary containing lists tracking different loss metrics across epochs
+    '''
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    classifier.to(device)
+    optimizer = optim.Adam(classifier.parameters(), lr=1e-5)
+
+    # TODO make a function equivalent to utils.get_class_weights
+    class_weights = {0: .9, 1: .1}
+
+    tune_label_counts, _ = utils.get_value_counts(tune_loader)
+    baseline = max(list(tune_label_counts.values())) / len(tune_dataset)
+
+    results = {'train_loss': [], 'tune_loss': [], 'train_acc': [], 'tune_acc': [],
+               'baseline': baseline, 'train_penalty': [], 'train_raw_loss': []}
+
+    train_sample_count = sum([utils.count_items_in_dataloader(dl) for dl in train_loaders])
+
+    try:
+        dummy_w = torch.nn.Parameter(torch.FloatTensor([1.0])).to(device)
+        best_tune_loss = None
+
+        for epoch in tqdm_notebook(range(num_epochs)):
+            train_correct = 0
+            train_loss = 0
+            train_penalty = 0
+            train_raw_loss = 0
+            for study_loader in train_loaders:
+                for batch in study_loader:
+                    expression, labels, ids = batch
+                    expression = expression.float().to(device)
```

Comment:
It shouldn't, `.to(device)` operates in place and returns the modified object so it could be called in either location. I imagine `float()` operates the same way from what I've seen of pytorch so far, but I'm not certain